source/common/common/token_bucket.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include "envoy/common/time.h"
+
+#include "common/common/utility.h"
+
+namespace Envoy {
+
+/**
+ * This class implements the token bucket algorithm (not thread-safe).
+ *
+ * https://en.wikipedia.org/wiki/Token_bucket
+ */
+class TokenBucket {
+public:
+  /**
+   * @param max_tokens supplies the maximun number of tokens in the bucket.
+   * @param fill_rate supplies the number of tokens that will return to the bucket on each second;
+   * Default is 1 token/second.
+   * @param time_source supplies the the time source and used to facilitate testing; Default is
+   * ProdMonotonicTimeSource.
+   */
+  TokenBucket(uint64_t max_tokens, double fill_rate = 1,
+              MonotonicTimeSource& time_source = ProdMonotonicTimeSource::instance_);
+
+  /**
+   * @param tokens supplies the number of tokens to be consumed. Default is 1.
+   * @return true if bucket is not empty, otherwise it returns false.
+   */
+  bool consume(uint64_t tokens = 1);
+
+private:
+  static constexpr double zero_time_ = 0.0;
```

Comment:
Where is this used?

Reply:
Still wondering?

Reply:
I forgot to delete it, that's all. Note that I'm still working on this code. Thanks for pointing it out though.